modules/host/network.go

```diff
@@ -5,21 +5,57 @@ import (
 	"sync/atomic"
 	"time"
 
+	"github.com/NebulousLabs/Sia/build"
 	"github.com/NebulousLabs/Sia/encoding"
 	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/types"
 )
 
+// threadedUpdateHostname periodically runs 'managedLearnHostname', which
+// checks if the host's hostname has changed, and makes an updated host
+// announcement if so.
+//
+// TODO: This thread doesn't actually have clean shutdown, because the sleep is
+// outside of of the resource lock.
+func (h *Host) threadedUpdateHostname() {
+	for {
+		h.resourceLock.RLock()
+		if h.closed {
+			// The host is closed, the goroutine can exit.
+			h.resourceLock.RUnlock()
+			break
+		}
+		h.managedLearnHostname()
+		h.resourceLock.RUnlock()
+
+		// Wait 30 minutes to check again. If the hostname is changing
+		// regularly (more than once a week), we want the host to be able to be
+		// seen as having 95% uptime. Every minute that the announcement is
+		// pointing to the wrong address is a minute of perceived downtime to
+		// the renters.
+		time.Sleep(time.Minute * 30)
+	}
+}
+
 // initNetworking performs actions like port forwarding, and gets the host
 // established on the network.
-func (h *Host) initNetworking(address string) error {
+func (h *Host) initNetworking(address string) (err error) {
 	// Create listener and set address.
-	var err error
 	h.listener, err = h.dependencies.listen("tcp", address)
 	if err != nil {
 		return err
 	}
-	h.netAddress = modules.NetAddress(h.listener.Addr().String())
+	h.mu.Lock()
+	h.port = modules.NetAddress(h.listener.Addr().String()).Port()
+	if build.Release == "testing" {
+		// TODO: I'm not really sure if this is causing errors because the
+		// tests are bad, or because the design is bad. But what seems to be
+		// happening is that the host is realizing it has an empty autoAddress,
+		// and this is somehow confusing the contractor. Bears further
+		// investigation.
+		h.autoAddress = modules.NetAddress(":" + h.port)
```

Comment:
you could mock the upnp dependency, I suppose, or the `myExternalIP` function.

The hostdb will add the host to `allHosts`, but it won't be able to scan it if the net address is empty, so it will never be added to `activeHosts`.